# Assignment 3: 
## Reproduce the analysis of the Hard to Employ program in NY

Reproduce the results in cell 2 of Table 2.1 (Ever employed in a CEO transitional job), and cell 10 (Convicted of a felony). 

- The second null hypothesis should be stated (for the "Convicted of a felony after 3 years" data).

- Both tests, Z and chi-sq, should be completed for the "Convicted of a felony after 3 years" data.

- The result of the test in term the rejection of the Null should be stated in all cases (for both tests and both for the original "Ever employed in a CEO transitional job" data and the "Convicted of a felony after 3 years data").


In [10]:
import os
import sys
import numpy as np
import pylab as pl

%pylab inline

Populating the interactive namespace from numpy and matplotlib


### Reproducing cell 2 > taken from Dr. Bianco's skeleton Ipython notebook:

## NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05
 $H_0: P_0 - P_1 \geq$ 0
 
 $H_a: P_0 - P_1 $< 0
 
 $\alpha$ = 0.05
 
#### this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above):
 $P_0=0.035, P_1=0.701$

In [22]:
alpha=0.05
# we like fractions better then percentages. as a rule of thumb, either use fractions or counts
P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

n_0 = 409
n_1 = 564

#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

we must assess the statistical significance


#### TWO OPTIONS z test, or chi-square test.

## Z TEST

#### the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the assumption of normality

In [14]:
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.421047276465


In [15]:
p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

** z score: how many standard deviation away from the population parameter is my statistic? **


$z=\frac{P_1-P_0}{\sigma}$

In [16]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)

20.7697865408


IMPORTANT!! note that this P in the bottom line of the table is not the p-value, but

p-value = 1-P

In [23]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


### Ever employed Z-test Result statement:
Looking at the Z chart, we set p-value (in our case: _p_2y_) using the P on the specific place. the _p_2y_ turned out to be much smaller than our significance threshold alpha = 0.05; that means it is way far out of the area of the null hypothesis, and for this reason can reject it.

## $\\chi^2$ TEST

In [18]:
from evalChisq import evalChisq

In [24]:
Ntot = 973 # a + b + c + d = tot

sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])

print (evalChisq(sample_values))

436.223462575


### Ever employed $\\chi^2$ TEST Result statement:
Looking at the $\\chi^2$ chart, using 1 degree of freedom and alpha = 0.05, the result that beyond it we could reject the Null hypothesis is 3.84. The $\\chi^2$ test, 436.22, turned out to be much larger than 3.84; that means it is way far out of the area of the null hypothesis, and for this reason can reject it.

__________

### Reproducing cell 10 > "Convicted of a felony after 3 years":

## NULL HYPOTHESIS: the % of former prisoners who were convivted of a felony is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05
 $H_0: P_0 - P_1 \geq$ 0
 
 $H_a: P_0 - P_1 $< 0
 
 $\alpha$ = 0.05
 
#### this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above):
 $P_0=0.117, P_1=0.100$

In [25]:
alpha=0.05
# we like fractions better then percentages. as a rule of thumb, either use fractions or counts
P2_0 = 11.7 * 0.01 
P2_1 = 10.0 * 0.01

if P2_0 - P2_1 >= 0:
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

n2_0 = 409
n2_1 = 568

#lets get the counts by multiplying by the sample size
Nt2_0 = P2_0 * n2_0
Nt2_1 = P2_1 * n2_1

the Null holds


## Z TEST

In [26]:
sp2 = (P2_0 * n2_0 + P2_1 * n2_1) / (n2_1 + n2_0)
print (sp2)

0.107116683726


In [27]:
p2 = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se2 = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

In [30]:
zscore2 = lambda p0, p1, s : (p0 - p1) / s
z_felony = zscore(P2_1, P2_0, se(p(P2_0, P2_1, n2_0, n2_1), n2_0, n2_1))
print (z_felony)

-0.84764020522


In [31]:
# use z chart to look for P
# p-value = 1-P
p_felony = 1 - 0.7995


def report_result(p2,a2):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p2,a2))
    if p2 < a2:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p2 < a2  else 'not rejected') )

    
report_result(p_felony, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


### Convicted of a felony Z-test Result statement:

Looking at the Z chart, we set p-value (in our case: _p_felony_) using the P on the specific place. the _p_felony_ turned out to be bigger than our significance threshold alpha = 0.05; that means it is within the area of the null hypothesis, and for this reason we cannot reject it.

## $\\chi^2$ TEST

In [35]:
# copied Dr. Bianco evalChisq() function

def evalChisq(values):
    '''Evaluates the chi sq from a contingency value
    Arguments:
    values: 2x2 array or list, the contingengy table
    '''
    if not (len(values.shape) == 2 and values.shape == (2,2)):
        print ("must pass a 2D array")
        return -1
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()

In [36]:
Ntot_felony = 977 # a + b + c + d = tot

sample_values_felony = np.array([[0.1 * 568, 0.9 * 568], [0.117 * 409, 0.883 * 409]])

sample_values_felony

array([[  56.8  ,  511.2  ],
       [  47.853,  361.147]])

In [37]:
print (evalChisq(sample_values_felony))

0.718493917505


### Convicted of a felony $\\chi^2$ TEST Result statement:
Looking at the $\\chi^2$ chart, using 1 degree of freedom and alpha = 0.05, the result that beyond it we could reject the Null hypothesis is 3.84. The $\\chi^2$ test, 0.718, turned out to be smaller than 3.84; that means it is within the area of the null hypothesis, and for this reason we cannot reject it.